Remove all instances of glBegin/glEnd usage and use batcher instead

ronaldfw opened the pull request:
This PR removes all instances of using glBegin/glEnd for drawing ("OpenGL immediate mode"). These calls are replaced by pushing primitives into a batcher (Batcher class) that is used for drawing all UI elements.

There are two batcher instances being used: One for drawing UI elements (member of GlCanvas), and one for drawing events on the timeline (Timer, callstack events; member of TimeGraph). The batcher in TimeGraph existed before. The one in GlCanvas is added here. It is pushed through all calls for drawing UI elements and these methods just add whatever elements they want to draw to the batcher. 

With the current batcher capabilities, some vertex transformations had to be implemented on the CPU (rotation for the toggle triangle, rotation for the rounded corners).

danielfenner commented on OrbitGl/Batcher.cpp:
```diff
@@ -67,6 +68,20 @@ void Batcher::AddShadedBox(Vec2 pos, Vec2 size, float z, Color color,
   AddBox(box, colors, picking_type, user_data);
 }
 
+void Batcher::AddTriangle(const Triangle& triangle, Color color,
```
Don't know how much of a difference it really makes but since this might be mildly preformance critical consider using a rvalue reference for the primitive. Same for Box and Line. Maybe this is rather a follow up optimization.

danielfenner replied:
Ahh, forget it. It is passed by reference down to the Block and there you need a copy anyway.

danielfenner commented on OrbitGl/Batcher.cpp:
```diff
@@ -95,4 +114,86 @@ void Batcher::GetBoxGradientColors(Color color, Color* colors) {
+
+//----------------------------------------------------------------------------
+void Batcher::DrawBoxBuffer(bool a_Picking) {
+  Block<Box, BoxBuffer::NUM_BOXES_PER_BLOCK>* boxBlock =
```
There are iterators in the BlockChain that might make this a bit more readable.

ronaldfw replied:
The iterators in BlockChain provide iterating over all elements in the blockchain, not iterating over blocks (which is what I need here). 

Btw, this code was moved from a different location and not really added in this PR. That's also why the naming is not according to our style guide. I'll change this here.

ronaldfw commented on OrbitGl/Track.cpp:
```diff
@@ -43,23 +43,50 @@ std::vector<Vec2> GetRoundedCornerMask(float radius, uint32_t num_sides) {
+  std::vector<Vec2> rotated_points = RotatePoints(points, rotation);
+
+  Vec3 position(pos[0], pos[1], z);
+  std::vector<Triangle> triangles;
```
Hmm, it seems not. I guess I had planned to do this differently and then never needed the triangles vector. I'll remove.

ronaldfw commented:
> Hi Ronald, thanks a lot for this substantial change. In general, looks really good to me. The only problem I see with having multiple batchers is picking. We encode the index of a textbox as a color when picking. When a user clicks, for the case of textboxes, we convert that color back to an index to get the textbox in the time_graph_'s batcher. This is a problem with multiple batchers as different boxes will have the same picking color, which is not correct and could potentially cause a crash. Maybe we could encode a "batcher id" in the picking color.

I think I need a more thorough explanation of this part. If my understanding is correct, then I think this can easily be resolved: All of the elements that are added to the "UI batcher" (the member of GlCanvas) should have picking type INVALID. Since they used glBegin/glEnd to be rendered, they were certainly not pickable with this mechanism before. This would exclude them from the cases in CaptureWindow::Pick.

ronaldfw commented:
> I think I need a more thorough explanation of this part. If my understanding is correct, then I think this can easily be resolved: All of the elements that are added to the "UI batcher" (the member of GlCanvas) should have picking type INVALID. Since they used glBegin/glEnd to be rendered, they were certainly not pickable with this mechanism before. This would exclude them from the cases in CaptureWindow::Pick.

Hmm, I guess some of them need to be of type PICKABLE, so the pick request is sent to the PickingManager.

ronaldfw commented:
@pierricgimmig I've updated the code to now handle picking correctly by adding, as discussed, a batcher id that is now also encoded in the picking color. I removed unused PickingID types INVALID and EVENT (EVENT was set in some places but never read). I am not sure I'm using PICKABLE correct in some cases. For elements drawn that do not have a text box, I am using this type, however there are some cases where this doesn't really make sense (and I think if these are looked up in the picking manager, they are ignored as we just get a nullptr). Not sure about all implications of this.
